### DISCUSSION TITLE
Getting unrecognized arguments error due to multiprocessing.resource_tracker for bundled cli application

### DISCUSSION THREAD

**Discussion by TobiWo** [pyinstaller/pyinstaller#8122] [Help] [answered]
Hi, I'm building the following cli application: https://github.com/TobiWo/eth-duties .

I'm using `SharedMemory` from `multiprocessing.shared_memory` to share objects between different processes. When I bundle my application I'm getting weird errors like:

```
eth-duties: error: unrecognized arguments: -B -S -E -s -c from multiprocessing.resource_tracker import main;main(5)
```

I interpret this as that the application gets started with `-B` etc. which I don't handover during startup. It feels like that these arguments coming from the resource_tracker which is pretty strange.

I already tested this [recipe](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Multiprocessing#required-changes-to-your-code) but it does not help.

I also saw this [discussion](https://github.com/orgs/pyinstaller/discussions/8113) which seems similar but I don't understand the answer. I don't use `sys.executable` but `Process` from the multiprocessing package see [here](https://github.com/TobiWo/eth-duties/blob/main/duties/rest/core/server.py). I want to start a uvicorn server and this behavior is the recommended way while running uvicorn as a side process.

I currently tested the bundled binary/executable on the following OS:

* Ubuntu 22.04 :white_check_mark: 
* Windows 10 :x: 
* Ubuntu 20.04 :x: 

I will test on the OS marked with :x:.

Any ideas?

**Comment by bwoodsend** [comment]
Is [multiprocessing.freeze_support()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.freeze_support) getting called?

**Reply by TobiWo** [reply]
As mentioned I tested the `freeze_support` recipe but it didn't work. In detail I added the command at the beginning of the main block as described. Also I did not test on Windows so I have the issue on Linux/Ubuntu. I will edit my initial post.

**Reply by rokm** [reply]
Can we see *how and where* you tried to call `multiprocessing.freeze_support`?

**Reply by rokm** [reply]
Because while the recipe (which is outdated, by the way) says you need to call it "straight after the if `__name__ == '__main__':` line of the main module." that is not necessarily true - you need to call it before you use any `multiprocessing` primitives (e.g., before all imports that might use `multiprocessing` primitives as part of their initialization).

**Reply by rokm** [reply, marked as answer]
And in your case, you clearly need to call it before you [import `cli.arguments`](https://github.com/TobiWo/eth-duties/blob/a3ded2b3e0e6ab3ca49027c4017489572db8ee0a/duties/main.py#L10), as that will try to parse the command-line arguments for your main application during its initialization. So you need to divert the program flow for `multiprocessing` subprocess before it reaches that point.

**Reply by TobiWo** [reply]
Ok, I get it. This helps hopefully. I will give it a try. Thank you :pray:

**Reply by TobiWo** [reply]
@rokm I just tested while adding the `freeze_support` right before [this line](https://github.com/TobiWo/eth-duties/blob/main/duties/cli/arguments.py#L303) and it seems to work. Thanks a ton again :pray:
